src/yunohost/app.py

```diff
@@ -459,12 +469,42 @@ def app_upgrade(app=[], url=None, file=None):
         elif app_dict["upgradable"] == "url_required":
             logger.warning(m18n.n('custom_app_url_required', app=app_instance_name))
             continue
-        elif app_dict["upgradable"] == "yes":
+        elif app_dict["upgradable"] == "yes" or force:
             manifest, extracted_app_folder = _fetch_app_from_git(app_instance_name)
         else:
             logger.success(m18n.n('app_already_up_to_date', app=app_instance_name))
             continue
 
+        # Manage upgrade type and avoid any upgrade if there are nothing to do
+        upgrade_type = "UNKNOWN"
+        # Get current_version and new version
+        app_new_version = manifest.get("version", "?")
+        app_current_version = app_dict.get("version", "?")
+
+        if manifest.get('integration', {}).get("upgrade_only_if_version_changes", None) is True:
+
+            # do only the upgrade if there are a change
+            if version.parse(app_current_version) >= version.parse(app_new_version) and not force:
+                logger.success(m18n.n('app_already_up_to_date', app=app_instance_name))
+                # Save update time
+                now = int(time.time())
+                app_setting(app_instance_name, 'update_time', now)
+                app_setting(app_instance_name, 'current_revision', manifest.get('remote', {}).get('revision', "?"))
+                continue
+            elif version.parse(app_current_version) > version.parse(app_new_version):
+                upgrade_type = "DOWNGRADE_FORCED"
+            elif app_current_version == app_new_version:
+                upgrade_type = "UPGRADE_FORCED"
+            elif "~ynh" in app_current_version and "~ynh" in app_new_version:
+                app_current_version_upstream, app_current_version_pkg = app_current_version.split("~ynh")
+                app_new_version_upstream, app_new_version_pkg = app_new_version.split("~ynh")
+                if app_current_version_upstream == app_new_version_upstream:
+                    upgrade_type = "UPGRADE_PACKAGE"
+                elif app_current_version_pkg == app_new_version_pkg:
+                    upgrade_type = "UPGRADE_APP"
+                else:
+                    upgrade_type = "UPGRADE_FULL"
```

Comment:
Uh it's not clear what happens in the remaining "else" case here ? Sounds like if the old or new app doesn't have an `~ynh` prefix, the upgrade_type is gonna stay to unknown which is a bit weird ?

Reply:
Yes, after reading a new time this code it's confusing. Now it should be better.